## Project 1, Data Preprocessing

### Project description

The project is to prepare a report for a bank’s loan division. I have to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.
The report intended to be considered when building a credit score for a potential customer. A credit score is used to evaluate the ability of a potential borrower to repay their loan.

### Table of Contents

* [Step 1. General Information](#Step-1.-General-Information) <br /> General info observation, file opened

* [Step 2. Data Preprocessing](#Step-2.-Data-Preprocessing) <br />preprocessing steps: handling missings, converting floats to integers, categorizing 

    * [2.1 Identifying and filling in missing values](#2.1-Identifying-and-filling-in-missing-values) <br />checking every column for missing values, investigating them, filling and omitting missings
    
        * [2.1.1 'children' column](#2.1.1-'children'-column) <br />missings found, some peculiarities here
        
        * [2.1.2 'days_employed' column](#2.1.2-'days_employed'-column) <br />missings investigated and the column ommited 
        
        * [2.1.3 'dob_years' column](#2.1.3-'dob_years'-column) <br />missings detected and processed
        
        * [2.1.4 'education' column](#2.1.4-'education'-column) <br />uppercase duplicates found
        
        * [2.1.5 'education_id' column](#2.1.5-'education_id'-column) <br />clear

        * [2.1.6 'family_status' column](#2.1.6-'family_status'-column) <br />clear

        * [2.1.7 'family_status_id' column](#2.1.7-'family_status_id'-column) <br />clear
        
        * [2.1.8 'gender' column](#2.1.8-'gender'-column) <br />missing value detected and processed       
        
        * [2.1.9 'income_type' column](#2.1.9-'income_type'-column) <br />underrepresented categories problem        
        
        * [2.1.10 'debt' column](#2.1.10-'debt'-column) <br />clear.     
        
        * [2.1.11 'total_income' column](#2.1.11-'total_income'-column)  <br />missings detected and processed            
        
        * [2.1.12 'purpose' column](#2.1.12-'purpose'-column) <br />clear        
        
    * [2.2 Replacing the real number data type with the integer type](#2.2-Replacing-the-real-number-data-type-with-the-integer-type) 
    
    * [2.3 Getting rid of duplicates](#2.3-Getting-rid-of-duplicates)   

        * [2.3.1 General](#2.3.1-General) <br />dropping duplicated rows for the whole dataframe     
        
        * [2.3.2 education](#2.3.2-education) <br />uppercase duplicates lowercased            
        
        * [2.3.3 income_type](#2.3.3-income_type) <br />logical duplication      
        
    * [2.4 Categorizing the data](#2.4-Categorizing-the-data)            
        
        * [2.4.1 categorization by purpose](#2.4.1-categorization-by-purpose)             
        
        * [2.4.2 categorization by income type](#2.4.2-categorization-by-income-type)            
        
* [Step 3. Answering questions](#Step-3.-Answering-questions)         
        
    * [3.1 Is there a connection between having kids and repaying a loan on time?](#3.1-Is-there-a-connection-between-having-kids-and-repaying-a-loan-on-time?)
    
    * [3.2 Is there a connection between marital status and repaying a loan on time?](#3.2-Is-there-a-connection-between-marital-status-and-repaying-a-loan-on-time?)  
    
    * [3.3 Is there a connection between income level and repaying a loan on time?](#3.3-Is-there-a-connection-between-income-level-and-repaying-a-loan-on-time?)
    
    * [3.4 How do different loan purposes affect on-time loan repayment?](#3.4-How-do-different-loan-purposes-affect-on-time-loan-repayment?)      
        
* [Step 4. General conclusion](#Step-4.-General-conclusion)

* [Project Readiness Checklist](#Project-Readiness-Checklist)


### Step 1. General Information

Dataframe legend:

1. children: the number of children in the family
2. days_employed: how long the customer has been working
3. dob_years: the customer’s age
4. education: the customer’s education level
5. education_id: identifier for the customer’s education
6. family_status: the customer’s marital status
7. family_status_id: identifier for the customer’s marital status
8. gender: the customer’s gender
9. income_type: the customer’s income type
10. debt: whether the customer has ever defaulted on a loan
11. total_income: monthly income
12. purpose: reason for taking out a loan


In [1]:
import pandas as pd
import numpy as np
credit_scoring = pd.read_csv('/datasets/credit_scoring_eng.csv')
credit_scoring.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [2]:
credit_scoring['debt'].value_counts()

0    19784
1     1741
Name: debt, dtype: int64

For the further work I assume that 1 stands for positive and 0 for negative answer to the question if they had ever defaulted a loan. Hope it's correct, it was not mentioned in the task description.  

In [3]:
credit_scoring.head(60)


children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42    bachelor's degree             0   
1          1   -4024.803754         36  secondary education             1   
2          0   -5623.422610         33  Secondary Education             1   
3          3   -4124.747207         32  secondary education             1   
4          0  340266.072047         53  secondary education             1   
5          0    -926.185831         27    bachelor's degree             0   
6          0   -2879.202052         43    bachelor's degree             0   
7          0    -152.779569         50  SECONDARY EDUCATION             1   
8          2   -6929.865299         35    BACHELOR'S DEGREE             0   
9          0   -2188.756445         41  secondary education             1   
10         2   -4171.483647         36    bachelor's degree             0   
11         0    -792.701887         40  secondary education             1   
12         0            NaN         65  secondary education             1   
13         0   -1846.641941         54         some college             2   
14         0   -1844.956182         56    bachelor's degree             0   
15         1    -972.364419         26  secondary education             1   
16         0   -1719.934226         35  secondary education             1   
17         0   -2369.999720         33    bachelor's degree             0   
18         0  400281.136913         53  secondary education             1   
19         0  -10038.818549         48  SECONDARY EDUCATION             1   
20         1   -1311.604166         36  secondary education             1   
21         1    -253.685166         33  secondary education             1   
22         1   -1766.644138         24  secondary education             1   
23         0    -272.981385         21    bachelor's degree             0   
24         1  338551.952911         57  secondary education             1   
25         0  363548.489348         67  secondary education             1   
26         0            NaN         41  secondary education             1   
27         0    -529.191635         28    bachelor's degree             0   
28         1    -717.274324         26    bachelor's degree             0   
29         0            NaN         63  secondary education             1   
30         1  335581.668515         62  secondary education             1   
31         0   -1682.083438         47    primary education             3   
32         0   -4649.910832         34  SECONDARY EDUCATION             1   
33         0   -1548.637544         48  secondary education             1   
34         0   -4488.067031         35  secondary education             1   
35         0  394021.072184         68  secondary education             1   
36         0    -176.216688         33  secondary education             1   
37         0   -6448.810860         43    BACHELOR'S DEGREE             0   
38         0    -597.881827         25    bachelor's degree             0   
39         0    -650.587796         31    bachelor's degree             0   
40         0   -1030.362606         30  secondary education             1   
41         0            NaN         50  secondary education             1   
42         0   -1257.496190         20         some college             2   
43         0   -4375.681384         43         some college             2   
44         1   -1362.041728         26  SECONDARY EDUCATION             1   
45         1   -1039.451741         49  secondary education             1   
46         2   -2262.712304         37  secondary education             1   
47         1   -2689.137274         33    bachelor's degree             0   
48         0   -3341.067886         45  secondary education             1   
49         0   -1181.443228         54  secondary education             1   
50         0  353731.432338         63  secondary education             1   


In [4]:
credit_scoring.tail(60)


children  days_employed  dob_years            education  education_id  \
21465         1    -294.700811         53  Secondary Education             1   
21466         0   -1029.052708         37  secondary education             1   
21467         0  338060.788569         59  secondary education             1   
21468         0    -785.453500         43    BACHELOR'S DEGREE             0   
21469         0   -2808.519596         43    bachelor's degree             0   
21470         3   -1555.901115         38    bachelor's degree             0   
21471         0   -6241.387809         59  secondary education             1   
21472         2   -2768.868055         39    bachelor's degree             0   
21473         0   -9945.187489         48  secondary education             1   
21474         1    -394.879647         33    bachelor's degree             0   
21475         0  365213.306266         52  secondary education             1   
21476         0   -2539.502794         52    bachelor's degree             0   
21477         0   -1628.432350         54  secondary education             1   
21478         1    -921.332705         28    Bachelor's Degree             0   
21479         0    -685.565994         26  secondary education             1   
21480         0   -8799.354717         45    bachelor's degree             0   
21481         0    -829.519627         30         some college             2   
21482         1    -865.081118         25  secondary education             1   
21483         2   -2917.035232         38    primary education             3   
21484         3    -968.718123         35  secondary education             1   
21485         0  363996.449108         59    bachelor's degree             0   
21486         0   -1614.022809         36  SECONDARY EDUCATION             1   
21487         0   -2376.789344         30  SECONDARY EDUCATION             1   
21488         0    -354.755760         35    bachelor's degree             0   
21489         2            NaN         47  Secondary Education             1   
21490         2   -1867.006757         32  secondary education             1   
21491        20    -173.954460         27  secondary education             1   
21492         0   -1618.645283         54  secondary education             1   
21493         1   -1214.858885         38  secondary education             1   
21494         0    -960.095747         39  secondary education             1   
21495         1            NaN         50  secondary education             1   
21496         1    -759.680851         31  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21498         1   -1330.627998         32  secondary education             1   
21499         0   -9929.015065         57  secondary education             1   
21500         0    -578.082757         26         some college             2   
21501         0  334343.096304         57  secondary education             1   
21502         1            NaN         42  secondary education             1   
21503         0   -3096.881131         58  secondary education             1   
21504         0  355235.728158         68  secondary education             1   
21505         0  338904.866406         53  secondary education             1   
21506         1   -1556.249906         33    BACHELOR'S DEGREE             0   
21507         1     -79.832064         32  secondary education             1   
21508         0  386497.714078         62  secondary education             1   
21509         0  362161.054124         59    bachelor's degree             0   
21510         2            NaN         28  secondary education             1   
21511         0    -612.569129         29    bachelor's degree             0   
21512         0    -165.377752         26    bachelor's degree             0   
21513         0   -1166.216789         35  secondary education             1   
21514  

In [5]:
credit_scoring.sample(60)

children  days_employed  dob_years            education  education_id  \
19346         1  332290.802045         59  secondary education             1   
12337         1   -1195.621516         45  secondary education             1   
8339          0   -2690.324007         27  secondary education             1   
13267         0  357762.286557         65  secondary education             1   
730           0   -2419.449434         52  secondary education             1   
1894          0    -595.559207         20  secondary education             1   
4386          0    -946.198260         20  secondary education             1   
2286          2   -4788.705354         48  secondary education             1   
14269         2    -481.075050         36  secondary education             1   
16055         1   -1082.054126         35  secondary education             1   
15030         0  383923.904129         57    bachelor's degree             0   
20313         1    -414.690877         31  secondary education             1   
10501         0    -338.127808         27         some college             2   
12590         0   -4175.855736         35    bachelor's degree             0   
375           0   -2491.666580         33  Secondary Education             1   
18625         2    -229.890507         30  secondary education             1   
1093          0  391127.337570         59  secondary education             1   
6468          1   -2103.355440         26  Secondary Education             1   
2729          0  332519.742268         59  secondary education             1   
11473         0    -514.210381         33  Secondary Education             1   
6376          0   -2792.710442         47    BACHELOR'S DEGREE             0   
12862         2    -819.649418         35  secondary education             1   
18762         0  369428.180859         45  secondary education             1   
2009          1   -5153.814640         33    Bachelor's Degree             0   
18550         0            NaN         58  SECONDARY EDUCATION             1   
13404         0            NaN         22  secondary education             1   
8821          0   -4392.591058         52  secondary education             1   
4962          0  367766.716364         55  secondary education             1   
342           0   -1821.341911         37  secondary education             1   
4248          2   -2191.261971         30  secondary education             1   
18768         0  365467.395074         63  secondary education             1   
14956         1   -2908.689791         44  secondary education             1   
4867          0    -512.161757         32  Secondary Education             1   
18640         0   -1311.094348         34  secondary education             1   
15197         0   -2548.989295         44  secondary education             1   
9461          0   -5562.311365         61  secondary education             1   
7585          1            NaN         27  Secondary Education             1   
9177          0     -83.670709         48  secondary education             1   
3148          0   -6018.146376         45  SECONDARY EDUCATION             1   
18318         0   -1171.295129         25    bachelor's degree             0   
8043          0    -768.326265         40    bachelor's degree             0   
17425         0  396783.883705         65  secondary education             1   
11247         0   -4479.089234         57  Secondary Education             1   
6426          0    -877.838402         33  secondary education             1   
6511          0    -457.476582         64  secondary education             1   
882           0    -250.969987         35    bachelor's degree             0   
8895          0   -3218.434996         31    primary education             3   
14105         0            NaN         63  secondary education             1   
6499          0   -1675.953605         40  secondary education             1   
19795  

Here I called .info(), .describe(), .head(), .tail(), .sample() functions to view some basic info on this dataset. While doing it I discovered that Jupyter shows tables as these nice looking scrollable frames if I request number of rows between 40 and 60 inclusively, so I'm calling 60 rows oftenly.   

I'd say here we have 8 categorical and 4 quantitative variables in 12 columns of 21524 rows. Some data is probably missing, noticeable ~2k difference in rows number for 'total_income' & 'days_employed' columns, at first sight the former one seem to contain meaningless data (investigating it below). 

Only 'days_employed' & 'total_income' columns are both of *float64* type.

Overall, I hope this is sufficient piece of data to make somewhat fair conclusions for the project. Let's see what's left after data preprocessing...

### Step 2. Data Preprocessing

#### 2.1 Identifying and filling in missing values

Here I'm checking columns for missing values, and replacing them when needed.

In [6]:
credit_scoring.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt   total_income  
count      21525.000000  21525.000000   19351.000000  
mean           0.972544      0.080883   26787.568355  
std            1.420324      0.272661   16475.450632  
min            0.000000      0.000000    3306.762000  
25%            0.000000      0.000000   16488.504500  
50%            0.000000      0.000000   23202.870000  
75%            1.000000      0.000000   32549.611000  
max            4.000000      1.000000  362496.645000

The expression above shows general information, but I feel like it belongs to this step, I'm referring to it below 
I think by the children, income, and education columns, I could even try to identify the region, 
by comparing it to the world data sorted by regions. 

##### 2.1.1 'children' column

In [7]:
credit_scoring['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

So, we have some outliers here. 
20 children is too much, worlds average is 2.5 as shown below. Most likely the data is distorted because of human input, but it also might be some program error, I have no origination story of this dataset to make clear conclusion why is it messed up so bad, don't think it actually look like what bank staff deal with regularly. So, if 20 is typo, it might both stand for 2 or 0. I think I can replace it with the median value, since it appears to be 0 as well, I calculated the median for the column values with the outliers excluded below.   

I'd guess that 47 negative values could stand for 1, just wrong sign. But I don't feel like it's typo, on standard keyboard layouts 1 and minus sign are not as close as 2 and 0. One could do this in couple of misclicks in Excel, or it could be misprocessed by some application.

In [8]:
f_data = pd.read_fwf('https://www.cia.gov/library/publications/the-world-factbook/rankorder/rawdata_2127.txt', names=('A', 'B', 'C'))
print ("World's median fertility per woman: {:.2f}".format(f_data['C'].median()))

World's median fertility per woman: 2.04


Not like it's super relevant here, but since I found this qute piece of data...
Discovered read_fwf function in pandas documentation, and fertility data by counrty in the CIA factbook.
I'm not completely sure how that data was gathered and calculated, but there is no country with 0 children in average in the list. Even though this is fertility rate per woman, and per-male or crude rates should be different, nowhere it is 0. So I may suggest that our sample is not representative in relation to a general population, but on the other hand, it should not be, I guess. 

In [9]:
children_temp = credit_scoring['children'].replace([20, -1], np.nan)
children_temp.value_counts()

0.0    14149
1.0     4818
2.0     2055
3.0      330
4.0       41
5.0        9
Name: children, dtype: int64

Here I defined the children_temp variable for storing children column with the outlying values excluded. Just to find it's median value to compare it to the median of the untouched 'children' column, in order to see how outlier-proof the .median() function actually is in our case. I'm also curious to compare it to the mean values, since I'm not firmly sure yet when it's better to use mean and when median for missing values imputation.

In [10]:
children_temp.median()

0.0

In [11]:
children_temp.mean()

0.4731800766283525

In [12]:
credit_scoring['children'].median()

0.0

In [13]:
credit_scoring['children'].mean()

0.5389082462253194

OK. Median values are similar both with and without outliers. So I'm replacing missing entries with zeroes.

In [14]:
credit_scoring['children'] = credit_scoring['children'].replace([20, -1], 0)
credit_scoring['children'].value_counts()

0    14272
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

Nice. 

##### 2.1.2 'days_employed' column

In [15]:
print ('For the days_employed column: ''Negative values', credit_scoring['days_employed'][credit_scoring['days_employed'] < 0].count(), '|', 'Positive values', credit_scoring['days_employed'][credit_scoring['days_employed'] > 0].count(), '|', 'Zeroes', credit_scoring['days_employed'][credit_scoring['days_employed'] == 0].count())

For the days_employed column: Negative values 15906 | Positive values 3445 | Zeroes 0


In [16]:
print ('Checking for missing values with isna() function:', credit_scoring['days_employed'].isna().sum(), '|', 'Checking non-missing values with notna() function:', credit_scoring['days_employed'].notna().sum())

Checking for missing values with isna() function: 2174 | Checking non-missing values with notna() function: 19351


In [17]:
credit_scoring[credit_scoring['days_employed'].isna() == False].sort_values(by = 'dob_years', ascending = False).head(60)

children  days_employed  dob_years            education  education_id  \
8880          0   -1678.969771         75  secondary education             1   
3460          0  344623.836105         74  secondary education             1   
11532         0   -6682.867814         74  secondary education             1   
4895          0  341528.126150         74    bachelor's degree             0   
19642         0  380150.387046         74  secondary education             1   
12317         0   -1729.632531         74  secondary education             1   
2557          0  372861.103965         74  secondary education             1   
15450         0   -3429.205485         73  secondary education             1   
229           0  336747.835284         73  secondary education             1   
5804          0  399563.882630         73  secondary education             1   
1826          0  368375.048770         73  Secondary Education             1   
20610         0  364291.049263         73  secondary education             1   
19385         0  372290.349044         73    primary education             3   
6406          0  352239.799751         73  secondary education             1   
2801          0  400008.553225         72  secondary education             1   
8765          0   -8615.516055         72    Bachelor's Degree             0   
10002         0  383159.004530         72  secondary education             1   
6221          0  389307.068999         72  secondary education             1   
11876         0  396541.413988         72    bachelor's degree             0   
20444         0  328728.720605         72  secondary education             1   
5461          0  372210.248480         72  secondary education             1   
8541          0  340928.726912         72  secondary education             1   
16184         0  399150.259251         72  secondary education             1   
14797         0  399413.111767         72  secondary education             1   
11704         0  348717.634761         72  secondary education             1   
4499          0  343510.590449         72    bachelor's degree             0   
1239          0  374093.630347         72  secondary education             1   
6099          0  335135.254672         72    primary education             3   
18568         0  348482.310733         72    bachelor's degree             0   
19077         0  347274.759207         72    bachelor's degree             0   
6516          0  -10093.858479         72    bachelor's degree             0   
16552         0  367378.973169         72  secondary education             1   
8728          0  375453.429975         72  secondary education             1   
4613          1  353765.642310         72  secondary education             1   
16867         0  342027.401238         72  secondary education             1   
12510         0  369733.465222         72  secondary education             1   
6393          0  382730.859065         72  secondary education             1   
4963          0  360639.907161         72    primary education             3   
3365          1  396057.535120         72  secondary education             1   
2446          0  355096.478464         72  secondary education             1   
2399          0   -1102.385647         72  secondary education             1   
19511         0  382422.007089         72  secondary education             1   
6649          0  345409.214065         72  secondary education             1   
21257         0  379985.925569         72  secondary education             1   
2454          1  391338.613127         72    primary education             3   
18932         0  352061.569533         71  SECONDARY EDUCATION             1   
1407          0  340176.401864         71  secondary education             1   
5326          0   -1504.924191         71  secondary education             1   
14873         0  359284.421236         71  secondary education             1   
5361   

In [18]:
credit_scoring.sort_values(by = 'days_employed', ascending = False).head(60)

children  days_employed  dob_years            education  education_id  \
6954          0  401755.400475         56  secondary education             1   
10006         0  401715.811749         69    bachelor's degree             0   
7664          1  401675.093434         61  secondary education             1   
2156          0  401674.466633         60  secondary education             1   
7794          0  401663.850046         61  secondary education             1   
4697          0  401635.032697         56  secondary education             1   
13420         0  401619.633298         63  Secondary Education             1   
17823         0  401614.475622         59  secondary education             1   
10991         0  401591.828457         56  secondary education             1   
8369          0  401590.452231         58  secondary education             1   
1184          0  401575.196728         55  secondary education             1   
4949          0  401573.905288         54  secondary education             1   
15192         0  401556.753550         68    Bachelor's Degree             0   
5716          0  401524.259153         57  secondary education             1   
10484         0  401517.276388         68  secondary education             1   
16237         0  401486.706746         64  SECONDARY EDUCATION             1   
15599         0  401458.877781         61  secondary education             1   
701           0  401446.446720         56    bachelor's degree             0   
14356         0  401440.834336         62    primary education             3   
5762          0  401381.736156         52  secondary education             1   
6904          0  401378.570761         67    bachelor's degree             0   
15285         0  401371.359788         59  SECONDARY EDUCATION             1   
1723          0  401369.483602         57  secondary education             1   
12578         0  401364.852882         63  secondary education             1   
7974          0  401345.926719         66  secondary education             1   
15760         0  401331.668947         70  secondary education             1   
4839          0  401321.964602         65  secondary education             1   
14177         0  401291.529739         57  secondary education             1   
12223         0  401271.363816         54  secondary education             1   
21378         0  401256.745090         59  secondary education             1   
7668          1  401226.800545         60  secondary education             1   
14983         0  401223.838667         57  secondary education             1   
19903         0  401215.459785         61  secondary education             1   
14955         0  401188.917385         50  secondary education             1   
8481          0  401161.158251         55  secondary education             1   
9897          0  401123.554203         61  Secondary Education             1   
20771         0  401113.119104         50  secondary education             1   
9802          0  401090.058209         60  SECONDARY EDUCATION             1   
5805          0  401063.804512         67    bachelor's degree             0   
19949         0  401035.292815         67    bachelor's degree             0   
8185          0  401021.500882         58  secondary education             1   
15972         1  401020.642572         53    primary education             3   
17112         0  400997.270898         57  secondary education             1   
434           0  400992.436570         66  secondary education             1   
14514         0  400992.375704          0  secondary education             1   
4227          0  400977.334538         61  secondary education             1   
7282          0  400961.909364         55    bachelor's degree             0   
11005         0  400954.247845         49  secondary education             1   
12059         1  400892.431139         58  secondary education             1   
9745   

In [19]:
credit_scoring[credit_scoring['days_employed'].isna() == False].sort_values(by = 'total_income').tail(60)

children  days_employed  dob_years            education  education_id  \
17192         0    -222.049875         42    bachelor's degree             0   
2468          0  -11094.366797         53    bachelor's degree             0   
11563         0   -1025.402943         64    bachelor's degree             0   
3055          0   -1347.392662         29    bachelor's degree             0   
6084          0  351717.389895         65    bachelor's degree             0   
17148         1    -956.435552         41  secondary education             1   
8264          0    -617.938882         46    bachelor's degree             0   
11652         2   -4458.032551         41    bachelor's degree             0   
16015         1   -3009.681981         50    bachelor's degree             0   
1845          3   -1687.668878         31    bachelor's degree             0   
15660         0  342783.405779         64    bachelor's degree             0   
9510          0   -3809.863000         30    bachelor's degree             0   
39            0    -650.587796         31    bachelor's degree             0   
13440         0    -925.966217         52    bachelor's degree             0   
3996          3   -2113.730191         43  secondary education             1   
11478         0   -1808.795346         46    Bachelor's Degree             0   
14213         1    -265.088994         32    bachelor's degree             0   
4740          1    -927.482101         37  SECONDARY EDUCATION             1   
21295         0    -352.048655         51    bachelor's degree             0   
1858          2    -933.916567         42    bachelor's degree             0   
6645          0    -961.485327         29         some college             2   
5044          2   -2042.286996         37  secondary education             1   
10199         0   -2020.387808         33    bachelor's degree             0   
16533         2   -4504.539720         40    bachelor's degree             0   
7893          1   -4436.709329         41  secondary education             1   
16189         1   -1148.917442         35    bachelor's degree             0   
19303         0   -4017.318674         30    bachelor's degree             0   
17177         0   -5110.659348         49    bachelor's degree             0   
19302         1    -526.862866         30  secondary education             1   
6483          1   -1428.471156         34    Bachelor's Degree             0   
3097          1   -7814.429301         49    bachelor's degree             0   
5289          0   -1636.482076         37    bachelor's degree             0   
13926         0   -6528.589597         44         some college             2   
8299          2   -2262.323747         39  secondary education             1   
10613         0   -1524.297686         30    bachelor's degree             0   
4129          0   -7992.968249         52    bachelor's degree             0   
9156          0     -65.163190         52    bachelor's degree             0   
18347         0   -5536.801152         35    bachelor's degree             0   
4415          0    -542.321356         45  secondary education             1   
15047         2   -1654.855987         39    bachelor's degree             0   
1590          1   -1283.436610         53    bachelor's degree             0   
19338         3   -1733.364682         34    bachelor's degree             0   
18766         2   -1019.604773         43    bachelor's degree             0   
10004         0   -1979.388318         58  SECONDARY EDUCATION             1   
9871          2   -3092.259823         56    bachelor's degree             0   
13090         1   -3531.320534         54    bachelor's degree             0   
15211         0   -2009.086175         34  secondary education             1   
19813         0   -2672.323106         50    bachelor's degree             0   
7447          0   -8626.777234         39  secondary education             1   
2224   

In [20]:
credit_scoring[credit_scoring['days_employed'] < 0].sample(60)

children  days_employed  dob_years            education  education_id  \
8892          1   -3094.004435         46    bachelor's degree             0   
12102         0   -4541.158702         61  secondary education             1   
7373          2    -199.586379         42  secondary education             1   
10223         0    -151.581945         30  secondary education             1   
14388         0   -4415.934458         36  secondary education             1   
17540         0   -1044.593125         33  secondary education             1   
19781         3    -812.872772         33  secondary education             1   
20363         0   -2710.486179         50  secondary education             1   
9757          0    -974.110593         26    bachelor's degree             0   
14964         1   -2931.708413         30         some college             2   
17819         2    -878.125564         31  secondary education             1   
14568         0   -3423.447872         47  secondary education             1   
3196          0    -155.990060         46  secondary education             1   
8268          0   -1877.764077         46  secondary education             1   
21413         1   -6812.490625         40  secondary education             1   
20100         1   -1352.078266         32  SECONDARY EDUCATION             1   
4407          1    -919.661824         33  SECONDARY EDUCATION             1   
4130          0   -5147.039344         31    bachelor's degree             0   
12007         0   -1302.805866         40  secondary education             1   
12799         0   -7345.453207         44  secondary education             1   
12517         0    -372.073863         33  secondary education             1   
13467         0    -958.261048         24  secondary education             1   
4269          0    -240.193647         21    bachelor's degree             0   
2010          2   -1015.443160         36  secondary education             1   
12498         1    -635.909060         38  secondary education             1   
11296         0     -85.279668         37  secondary education             1   
12581         0   -1029.702347         55  secondary education             1   
2736          1    -232.630156         29  secondary education             1   
20988         0   -8058.043609         63  secondary education             1   
12116         0   -1752.594904         48  secondary education             1   
2376          0  -11862.758583         50  secondary education             1   
9867          0    -620.944772         26    bachelor's degree             0   
3803          1   -1509.982079         30  secondary education             1   
11127         2    -200.407954         41  secondary education             1   
15957         2   -4450.200886         35  secondary education             1   
410           0   -1776.594624         47    bachelor's degree             0   
70            1   -2802.226671         28    BACHELOR'S DEGREE             0   
20326         0    -458.706771         27    bachelor's degree             0   
5852          0   -3500.310848         40  secondary education             1   
2200          0   -3221.102349         55  secondary education             1   
12773         1   -2722.984427         39  secondary education             1   
5494          1   -1077.235702         30  secondary education             1   
17302         0    -142.222862         37  secondary education             1   
15191         0   -1157.009307         48  secondary education             1   
3650          0   -3876.154495         27  secondary education             1   
7284          0   -1333.025172         27  secondary education             1   
8024          0   -1429.442767         40    bachelor's degree             0   
8438          0   -1946.456093         35  SECONDARY EDUCATION             1   
12816         0   -2302.951113         30  secondary education             1   
16804  

In [21]:
credit_scoring[credit_scoring['days_employed'] > 0].sample(60)

children  days_employed  dob_years            education  education_id  \
8425          0  393964.840997         59  secondary education             1   
15346         0  374498.198822         58  secondary education             1   
935           0  373786.883727         60  secondary education             1   
9933          0  364967.359939         62    BACHELOR'S DEGREE             0   
7320          0  350601.295458         63  secondary education             1   
13125         0  377654.515876         54  secondary education             1   
20940         0  336848.368960         68  secondary education             1   
14302         0  356425.151366         54  secondary education             1   
16680         0  346458.191991         55  secondary education             1   
15782         0  398749.827786         54    bachelor's degree             0   
14409         0  333485.777074         60  secondary education             1   
4958          1  335559.216555         60  secondary education             1   
17547         0  375933.745908         53  secondary education             1   
1960          0  397812.163127         55  secondary education             1   
4893          0  343879.913091         54  secondary education             1   
9525          0  353413.166287         58  secondary education             1   
7490          0  362728.793964         59  secondary education             1   
10055         0  340850.240165         53  secondary education             1   
4403          0  352100.204966         58  secondary education             1   
15292         0  332390.428408         52  secondary education             1   
4333          1  348950.004171         55    Bachelor's Degree             0   
2801          0  400008.553225         72  secondary education             1   
3503          0  366501.202213         69  secondary education             1   
2031          0  358513.655290         66  secondary education             1   
4207          0  393154.692592         52  Secondary Education             1   
19395         0  361819.105096         54  secondary education             1   
18699         0  368556.010382         34  secondary education             1   
14370         0  370195.978096         64    bachelor's degree             0   
9267          0  331753.476757         67  SECONDARY EDUCATION             1   
12164         0  378174.187808         64  secondary education             1   
2753          0  332207.706723         58  secondary education             1   
4963          0  360639.907161         72    primary education             3   
17864         0  371113.904423         57  secondary education             1   
7543          0  346584.672744         59  secondary education             1   
3375          0  398885.186243         55  secondary education             1   
1731          0  364605.652117         62    bachelor's degree             0   
8931          0  354409.203887         62    bachelor's degree             0   
15457         0  384199.649161         51  secondary education             1   
4839          0  401321.964602         65  secondary education             1   
1638          0  354399.017910         59  secondary education             1   
12499         0  369403.970478         61  Secondary Education             1   
8270          0  379568.387414         65  secondary education             1   
12550         0  330226.301953         59  secondary education             1   
21053         0  392388.182360         59    bachelor's degree             0   
10716         0  391379.070727         61  secondary education             1   
9453          1  358750.368550         60  secondary education             1   
19622         0  400000.732466         63  secondary education             1   
1326          1  363065.428387         62    bachelor's degree             0   
9355          0  374762.301257         58  secondary education             1   
1761   

In [22]:
print ('retirees with total_days value greater than 0:', credit_scoring['days_employed'][(credit_scoring['income_type'] == 'retiree') & (credit_scoring['days_employed'] > 0)].count(), 'total number of retiree entries:', credit_scoring['days_employed'][credit_scoring['income_type'] == 'retiree'].count())

retirees with total_days value greater than 0: 3443 total number of retiree entries: 3443


Here I looked at different arrangements of the table trying to spot important details, actually I did many more rearrangements, just left several for examples. 

The 'days_employed' column seem to contain absolutely irrelevant data: there are 15906 negative values in the column, 3445 remaining are postitives ranging from 328728.72060451825 to 401755.40047533 - so they seem to be meaningless as well, if the 'days_employed' column values represented number of days indeed, we could find that it ranges somewhat between ~900 and ~1100 years. It's not indicated in the project task, what currency we're talking about, neither it's mentioned if our dataset contains info about human beings with average lifespan, but here I'm assuming so.

What is peculiar about the 'days_employed' column, is that the very same rows are missing in the 'total_income' column. 
Almost all retirees have positive nubmer in 'days_employed' (and several NaNs). It might be representing something completely different than number of days, like current bank account balance for instance, which is very unlikely beacuse of such a narrow range, thouht the balance could have negative values as well. Or it might be a result of some mistake, miscalculation or misprocessing of data, the data in the column does not seem to be completely random, but if it's actual meaning recovery is possible, it requires more informstion. 

I may conclude, this data is missing completely at random.

Don't think we have a reliable way to replace values with something meaningful in this column, so let's ommit it for the further steps. Right in the cell below I'm saving credit_scoring df with the 'days_employed' dropped. 

In [23]:
credit_scoring = credit_scoring.drop(columns=['days_employed'])
credit_scoring

children  dob_years            education  education_id  \
0             1         42    bachelor's degree             0   
1             1         36  secondary education             1   
2             0         33  Secondary Education             1   
3             3         32  secondary education             1   
4             0         53  secondary education             1   
...         ...        ...                  ...           ...   
21520         1         43  secondary education             1   
21521         0         67  secondary education             1   
21522         1         38  secondary education             1   
21523         3         38  secondary education             1   
21524         2         40  secondary education             1   

           family_status  family_status_id gender income_type  debt  \
0                married                 0      F    employee     0   
1                married                 0      F    employee     0   
2                married                 0      M    employee     0   
3                married                 0      M    employee     0   
4      civil partnership                 1      F     retiree     0   
...                  ...               ...    ...         ...   ...   
21520  civil partnership                 1      F    business     0   
21521            married                 0      F     retiree     0   
21522  civil partnership                 1      M    employee     1   
21523            married                 0      M    employee     1   
21524            married                 0      F    employee     0   

       total_income                  purpose  
0         40620.102    purchase of the house  
1         17932.802             car purchase  
2         23341.752    purchase of the house  
3         42820.568  supplementary education  
4         25378.572        to have a wedding  
...             ...                      ...  
21520     35966.698     housing transactions  
21521     24959.969        purchase of a car  
21522     14347.610                 property  
21523     39054.888        buying my own car  
21524     13127.587             to buy a car  

[21525 rows x 11 columns]

OK

##### 2.1.3 'dob_years' column

In [24]:
credit_scoring['dob_years'].isnull().sum()

0

In [25]:
credit_scoring['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Just inspected it visually. 101 zeroes, missing values for age column. 

In [26]:

credit_scoring['dob_years'].value_counts().count()

58

In [27]:
credit_scoring['dob_years'][credit_scoring['dob_years'] < 0].count()

0

This time no neagative values. Ah, wasted chance to seek for metaphysical explaination for negative age. 

In [28]:
credit_scoring['dob_years'][credit_scoring['dob_years'] < 0].count()

0

In [29]:
credit_scoring['dob_years'][credit_scoring['dob_years'] > 75].count()

0

In [30]:
credit_scoring['dob_years'][credit_scoring['dob_years'] < 19].count()

101

101 zero values. OK, let's replace it with the median value. I'm still trying to figure out when it's better to use median, and when mean, seems like it depends on what kind of data it is. Here I don't see extreme outliers, but I'd still prefer median, it is a whole number already, unlike mean, which I intentionally rounded to the nearest integer in the code. Anyway, for this particular case diffirence between mean and median is not that big. 

In [31]:
print ('median age:', credit_scoring['dob_years'].median(), '|', 'mean age:', round(credit_scoring['dob_years'].mean()))

median age: 42.0 | mean age: 43


In [32]:
credit_scoring['dob_years'] = credit_scoring['dob_years'].replace(0, 42)
credit_scoring['dob_years'][credit_scoring['dob_years'] == 0].count()

0

Done.

In [33]:
credit_scoring['dob_years'][credit_scoring['dob_years'] < 19].count()

0

Normal age range now.

##### 2.1.4 'education' column

In [34]:
credit_scoring['education'].isna().sum()

0

In [35]:
credit_scoring['education'].unique()

array(["bachelor's degree", 'secondary education', 'Secondary Education',
       'SECONDARY EDUCATION', "BACHELOR'S DEGREE", 'some college',
       'primary education', "Bachelor's Degree", 'SOME COLLEGE',
       'Some College', 'PRIMARY EDUCATION', 'Primary Education',
       'Graduate Degree', 'GRADUATE DEGREE', 'graduate degree'],
      dtype=object)

In [36]:
credit_scoring['education'].value_counts()

secondary education    13750
bachelor's degree       4718
SECONDARY EDUCATION      772
Secondary Education      711
some college             668
BACHELOR'S DEGREE        274
Bachelor's Degree        268
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

No missing values. Only uppercase duplicates, handling them in the corresponding section below.

##### 2.1.5 'education_id' column

In [37]:
credit_scoring['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

OK

##### 2.1.6 'family_status' column

In [38]:
print (credit_scoring['family_status'].unique(),  '\n' 'missing values by isna() function:', credit_scoring['family_status'].isna().sum())

['married' 'civil partnership' 'widow / widower' 'divorced' 'unmarried'] 
missing values by isna() function: 0


In [39]:
credit_scoring['family_status'].value_counts()

married              12380
civil partnership     4177
unmarried             2813
divorced              1195
widow / widower        960
Name: family_status, dtype: int64

OK

##### 2.1.7 'family_status_id' column

In [40]:
print (credit_scoring['family_status_id'].unique(), credit_scoring['family_status_id'].isna().sum())

[0 1 2 3 4] 0


In [41]:
credit_scoring['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

OK

##### 2.1.8 'gender' column

In [42]:
print (credit_scoring['gender'].unique(), credit_scoring['gender'].isna().sum(), credit_scoring['gender'].isnull().sum())

['F' 'M' 'XNA'] 0 0


We only have two genders in the dataframe...  'XNA' looks like a missing value marker. 

In [43]:
credit_scoring['gender'][credit_scoring['gender'] == 'XNA'].count()

1

Let's just get rid of it using .drop() method.

In [44]:
credit_scoring = credit_scoring.drop(credit_scoring[credit_scoring.gender == 'XNA'].index)
credit_scoring['gender'].unique()

array(['F', 'M'], dtype=object)

Done

##### 2.1.9 'income_type' column

In [45]:
credit_scoring['income_type'].unique()

array(['employee', 'retiree', 'business', 'civil servant', 'unemployed',
       'entrepreneur', 'student', 'paternity / maternity leave'],
      dtype=object)

In [46]:
credit_scoring['income_type'].value_counts()

employee                       11119
business                        5084
retiree                         3856
civil servant                   1459
entrepreneur                       2
unemployed                         2
student                            1
paternity / maternity leave        1
Name: income_type, dtype: int64

OK, but some categories are underrepresented for making any staistical conclusions. So that we could get rid of 'unemployed', 'student' (Why would I even consider 'student' as an income type? I won't.), 'paternity / maternity leave' for insufficient representation. (With the enterpreneurs I'm dealing in the duplicates section below) 

In [47]:
credit_scoring = credit_scoring.drop(credit_scoring[credit_scoring.income_type == 'unemployed'].index)
credit_scoring = credit_scoring.drop(credit_scoring[credit_scoring.income_type == 'student'].index)
credit_scoring = credit_scoring.drop(credit_scoring[credit_scoring.income_type == 'paternity / maternity leave'].index)

credit_scoring['income_type'].value_counts()

employee         11119
business          5084
retiree           3856
civil servant     1459
entrepreneur         2
Name: income_type, dtype: int64

Done.

##### 2.1.10 'debt' column

In [48]:
print (credit_scoring['debt'].value_counts(), '\n' 'isnull():', credit_scoring['debt'].isnull().sum())

0    19781
1     1739
Name: debt, dtype: int64 
isnull(): 0


OK. 

##### 2.1.11 'total_income' column

In [49]:
print ('missing values:', credit_scoring['total_income'].isna().sum(), '\n' 'total entries:', credit_scoring['total_income'].count())

missing values: 2174 
total entries: 19346


In [50]:
credit_scoring['total_income'][credit_scoring['total_income'] <= 0].count()

0

At least no negative values, no even zeroes actually.

In [51]:
credit_scoring['total_income'].max()

362496.645

In [52]:
credit_scoring[credit_scoring['total_income'].isnull() == True].tail(50)

children  dob_years            education  education_id  \
21000         2         35  secondary education             1   
21031         0         56  secondary education             1   
21032         0         60  secondary education             1   
21057         0         62  secondary education             1   
21060         0         23  secondary education             1   
21070         0         59    bachelor's degree             0   
21077         1         41  secondary education             1   
21085         0         48  secondary education             1   
21090         0         41         some college             2   
21102         0         60  secondary education             1   
21118         0         40  secondary education             1   
21124         0         66  secondary education             1   
21126         0         39  secondary education             1   
21132         0         47  secondary education             1   
21146         0         69  secondary education             1   
21152         1         40  secondary education             1   
21177         0         27  secondary education             1   
21182         0         28    bachelor's degree             0   
21206         1         34  secondary education             1   
21211         0         48  secondary education             1   
21226         3         38  secondary education             1   
21242         1         33    bachelor's degree             0   
21251         0         57         some college             2   
21258         1         40  secondary education             1   
21268         1         44    bachelor's degree             0   
21271         2         42  secondary education             1   
21281         1         30    bachelor's degree             0   
21294         0         45    BACHELOR'S DEGREE             0   
21300         2         45  Secondary Education             1   
21305         0         59  secondary education             1   
21308         0         48  secondary education             1   
21311         0         49  secondary education             1   
21321         0         56  Secondary Education             1   
21350         0         21  secondary education             1   
21364         0         50  secondary education             1   
21369         2         42  secondary education             1   
21390         0         53  secondary education             1   
21391         0         52  secondary education             1   
21407         1         36  secondary education             1   
21414         0         65  secondary education             1   
21415         0         54  secondary education             1   
21423         0         63  secondary education             1   
21426         0         49  secondary education             1   
21432         1         38         some college             2   
21463         1         35    bachelor's degree             0   
21489         2         47  Secondary Education             1   
21495         1         50  secondary education             1   
21497         0         48    BACHELOR'S DEGREE             0   
21502         1         42  secondary education             1   
21510         2         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
21000            married                 0      M       employee     0   
21031  civil partnership                 1      F        retiree     0   
21032            married                 0      F        retiree     0   
21057  civil partnership                 1      M       employee     0   
21060          unmarried                 4      F       employee     0   
21070            married                 0      F        retiree     0   
21077            married                 0      F       business     0   
21085  civil partnership                 1      F       employee     0   
21090            married 

In [53]:
credit_scoring[credit_scoring['total_income'].isnull() == True].head(50)

children  dob_years            education  education_id  \
12          0         65  secondary education             1   
26          0         41  secondary education             1   
29          0         63  secondary education             1   
41          0         50  secondary education             1   
55          0         54  secondary education             1   
65          0         21  secondary education             1   
67          0         52    bachelor's degree             0   
72          1         32    bachelor's degree             0   
82          2         50    bachelor's degree             0   
83          0         52  secondary education             1   
90          2         35    bachelor's degree             0   
94          1         34    bachelor's degree             0   
96          0         44  SECONDARY EDUCATION             1   
97          0         47    bachelor's degree             0   
120         0         46  secondary education             1   
121         0         29    bachelor's degree             0   
135         0         27  secondary education             1   
141         0         39  secondary education             1   
145         0         62  secondary education             1   
174         0         55    bachelor's degree             0   
181         0         26  secondary education             1   
189         1         30  secondary education             1   
205         1         31    bachelor's degree             0   
220         1         23         some college             2   
241         0         47  secondary education             1   
242         0         58  secondary education             1   
247         1         60    bachelor's degree             0   
250         0         54    bachelor's degree             0   
264         2         40  secondary education             1   
278         1         23  Secondary Education             1   
279         0         44  Secondary Education             1   
280         0         66  secondary education             1   
312         1         33  secondary education             1   
317         0         21    bachelor's degree             0   
320         0         63  secondary education             1   
321         1         34    bachelor's degree             0   
322         0         56  SECONDARY EDUCATION             1   
328         0         69  secondary education             1   
360         0         27  secondary education             1   
361         0         59  secondary education             1   
364         0         57  secondary education             1   
376         0         23         Some College             2   
383         0         55    bachelor's degree             0   
389         1         31  SECONDARY EDUCATION             1   
415         0         57    bachelor's degree             0   
416         0         58  secondary education             1   
421         0         45  secondary education             1   
444         0         53  secondary education             1   
458         0         48  secondary education             1   
466         0         56  secondary education             1   

         family_status  family_status_id gender    income_type  debt  \
12   civil partnership                 1      M        retiree     0   
26             married                 0      M  civil servant     0   
29           unmarried                 4      F        retiree     0   
41             married                 0      F  civil servant     0   
55   civil partnership                 1      F        retiree     1   
65           unmarried                 4      M       business     0   
67             married                 0      F        retiree     0   
72             married                 0      M  civil servant     0   
82             married                 0      F       employee     0   
83             married                 0      M       employee     0   

Let's take a look at rows with missing total_income data. All the same rows where days_employed were missing. 
What is peculiar about *days_employed* & *total_income* columns, is that they have same number of rows (19351), I did't find any other dependencies, looks like this data is missing completely at random.
Generally, total_income values look legit, but we have 2174 NaN values here.

In [54]:
print ('total_income median:', credit_scoring['total_income'].median(), '|', 'total_income mean:', credit_scoring['total_income'].mean())

total_income median: 23203.328 | total_income mean: 26789.375414245842


In [55]:
credit_scoring['total_income'].describe() 

count     19346.000000
mean      26789.375414
std       16476.301719
min        3306.762000
25%       16493.600500
50%       23203.328000
75%       32550.337000
max      362496.645000
Name: total_income, dtype: float64

Ah, OK, 50 percentile stands for median

In [56]:
credit_scoring['total_income'][credit_scoring['total_income'] > 300000].count()

2

I see some outliers here, and they're legit values, not outstanding too far, but still may affect the mean value significantly. So I'm going for median here as well. 

In [57]:
credit_scoring['total_income'] = credit_scoring['total_income'].fillna(23203.786)
credit_scoring['total_income'].isna().sum()

0

Finally.

##### 2.1.12 'purpose' column

In [58]:
credit_scoring['purpose'].isna().sum()

0

In [59]:
credit_scoring['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

In [60]:
credit_scoring['purpose'].value_counts()

wedding ceremony                            797
having a wedding                            777
to have a wedding                           774
real estate transactions                    676
buy commercial real estate                  664
housing transactions                        653
buying property for renting out             652
transactions with commercial real estate    651
housing                                     647
purchase of the house                       647
purchase of the house for my family         641
property                                    634
construction of own property                634
transactions with my real estate            630
building a real estate                      626
buy real estate                             623
building a property                         620
purchase of my own house                    620
housing renovation                          611
buy residential real estate                 607
buying my own car                       

Looks OK. Requires some categorization. 

#### 2.2 Replacing the real number data type with the integer type

In [61]:
credit_scoring.dtypes

children              int64
dob_years             int64
education            object
education_id          int64
family_status        object
family_status_id      int64
gender               object
income_type          object
debt                  int64
total_income        float64
purpose              object
dtype: object

Of what's left till this point I only have total_income values represented as a float64 type. Guess they would do better job for us as an integers. Here I'm using astype() method with the 'int' string argument standing for integer. When converting from float to integer numbers get truncated, but digits after decimal point are not very significant for us here, and what is more important, converting floats to integers increases precision of the further calculations. Also, in my opinion, converting to integers increases human readability.

In [62]:
credit_scoring['total_income'] = credit_scoring['total_income'].astype(int)
credit_scoring['total_income'].dtypes

dtype('int64')

In [63]:
credit_scoring['total_income'].sample(5)

3255     32759
12504    17156
15509    27493
16327    13800
2897      8203
Name: total_income, dtype: int64

Great.

#### 2.3 Getting rid of duplicates

######  2.3.1 General

In [64]:
credit_scoring.shape[0]

21520

just to remind number of rows

In [65]:
credit_scoring.duplicated().sum()

55

Checked for duplicates. Dropping duplicates and checking the result:

In [66]:
credit_scoring = credit_scoring.drop_duplicates().reset_index(drop = True)
credit_scoring.duplicated().sum()

0

OK. Used duplicated() method to see the duplicated rows of the dataframe, and the drop_duplicates() method to delete them. I also had to attach .reset_index(drop = True) to it, to make sure the old index is not kept and the new on is present. 

######  2.3.2 education

In [67]:
credit_scoring['education'].value_counts()

secondary education    13703
bachelor's degree       4709
SECONDARY EDUCATION      771
Secondary Education      711
some college             667
BACHELOR'S DEGREE        273
Bachelor's Degree        267
primary education        250
Some College              47
SOME COLLEGE              29
PRIMARY EDUCATION         17
Primary Education         15
graduate degree            4
Graduate Degree            1
GRADUATE DEGREE            1
Name: education, dtype: int64

This looks bad. Duplicate upper and lowercase values. Only 5 actually unique categories here. Almost certainly that's human input error, like unintentionally pressed capslock key, far less likely it might be a result of software misprocessing, and even hardware problem can not be completely excluded as possible reason (like broken keyboard). 

In [68]:
credit_scoring['education'] = credit_scoring['education'].str.lower()
credit_scoring['education'].unique()

array(["bachelor's degree", 'secondary education', 'some college',
       'primary education', 'graduate degree'], dtype=object)

In [69]:
credit_scoring['education'].value_counts()

secondary education    15185
bachelor's degree       5249
some college             743
primary education        282
graduate degree            6
Name: education, dtype: int64

Much better :)
OK, but graduates are totally underrepresented here. It's not even worth of merging into common tertiary education group with bachelors and college graduates. Let's proceed with it as is, but keeping in mind that groups representation distributed extremely unevenly, so that groups comparison might not actually be valid, aspecially for the 'graduate degree' group with absolutely insignificant representation of only 6 entries. 

######  2.3.3 income_type

As I noticed while looked for missing values, in the 'income_type' column we have sort of logical duplication, and I think it's OK to merge the 'entrepreneur' with the 'business' for this column. The word entrepreneur seem more aproppriate for the income type description amongst the other descriptive values here. 

In [70]:
credit_scoring['income_type'].mask(credit_scoring['income_type'] == 'business', 'entrepreneur', inplace=True)
credit_scoring['income_type'].value_counts()

employee         11090
entrepreneur      5081
retiree           3837
civil servant     1457
Name: income_type, dtype: int64

OK. logical duplicates in 'purpose' I will handle with the categorizing right below.

#### 2.4 Categorizing the data

Here I'm categorising the data by the purpose and by the income type because I need these categories to answer questions in the third step.  

###### 2.4.1 categorization by purpose 

In [71]:
credit_scoring['purpose'].value_counts()

wedding ceremony                            793
having a wedding                            772
to have a wedding                           769
real estate transactions                    675
buy commercial real estate                  662
housing transactions                        652
buying property for renting out             651
transactions with commercial real estate    650
housing                                     646
purchase of the house                       646
purchase of the house for my family         638
construction of own property                634
property                                    633
transactions with my real estate            627
building a real estate                      625
purchase of my own house                    620
buy real estate                             620
building a property                         619
buy residential real estate                 606
housing renovation                          606
buying my own car                       

In [72]:
credit_scoring['purpose'].value_counts().count()

38

In [73]:
credit_scoring['purpose'].unique()

array(['purchase of the house', 'car purchase', 'supplementary education',
       'to have a wedding', 'housing transactions', 'education',
       'having a wedding', 'purchase of the house for my family',
       'buy real estate', 'buy commercial real estate',
       'buy residential real estate', 'construction of own property',
       'property', 'building a property', 'buying a second-hand car',
       'buying my own car', 'transactions with commercial real estate',
       'building a real estate', 'housing',
       'transactions with my real estate', 'cars', 'to become educated',
       'second-hand car purchase', 'getting an education', 'car',
       'wedding ceremony', 'to get a supplementary education',
       'purchase of my own house', 'real estate transactions',
       'getting higher education', 'to own a car', 'purchase of a car',
       'profile education', 'university education',
       'buying property for renting out', 'to buy a car',
       'housing renovation', 'going

Played with stemming a bit here. I won't use for this task. Don't see any practical reason to use stemming or lemmatisation here. Looking at the array of the unique values I manually composed these dictionaries literally in one minute. Here we have 4 categories and 10 keywords.

vehicle = 'car'<br />
estate = 'estate', 'housing', 'property', 'house'<br />
education = 'education', 'educated', 'university'<br />
wedding = 'wedding'<br />

In [74]:
def purpose_category(row):
    entry = row['purpose']
    
    try:
    
        for word in entry:
            if 'car' in entry:
                return 'vehicle'
            elif 'estate' in entry:
                return 'estate' 
            elif 'housing' in entry:
                return 'estate'
            elif 'housing' in entry:
                return 'estate'
            elif 'property' in entry:
                return 'estate'
            elif 'house' in entry:
                return 'estate'
            elif 'education' in entry:
                return 'education'
            elif 'educated' in entry:
                return 'education'
            elif 'university' in entry:
                return 'education'
            elif 'wedding' in entry:
                return 'wedding'
    
    except: 
        
        print ('Oops! I messed it up ^______^')
        
        
credit_scoring['purpose_cat'] = credit_scoring.apply(purpose_category, axis = 1)

credit_scoring.sample(60)

children  dob_years            education  education_id  \
8592          1         32    bachelor's degree             0   
15326         1         40         some college             2   
9309          1         27    bachelor's degree             0   
9861          0         35    bachelor's degree             0   
5081          0         63  secondary education             1   
926           0         57  secondary education             1   
16405         0         41    bachelor's degree             0   
20310         0         36  secondary education             1   
20833         2         40  secondary education             1   
1233          0         27    bachelor's degree             0   
7629          0         59  secondary education             1   
12050         1         51  secondary education             1   
6028          0         24    bachelor's degree             0   
21369         0         66    bachelor's degree             0   
15543         0         38  secondary education             1   
19670         1         38  secondary education             1   
17817         0         55  secondary education             1   
9481          1         26  secondary education             1   
10896         0         40  secondary education             1   
633           0         26    bachelor's degree             0   
18512         1         39  secondary education             1   
4348          0         42  secondary education             1   
20844         0         60  secondary education             1   
18858         0         42  secondary education             1   
13663         0         31  secondary education             1   
7552          0         38    bachelor's degree             0   
503           1         30    bachelor's degree             0   
2250          2         36  secondary education             1   
11079         0         27  secondary education             1   
889           1         30  secondary education             1   
6728          0         39    bachelor's degree             0   
1305          0         51  secondary education             1   
1497          1         35    bachelor's degree             0   
258           1         47  secondary education             1   
6087          0         35  secondary education             1   
17605         1         38         some college             2   
18209         1         43  secondary education             1   
14497         0         57  secondary education             1   
15970         0         59  secondary education             1   
4459          0         40  secondary education             1   
21025         1         38    bachelor's degree             0   
7355          0         61  secondary education             1   
2598          0         26  secondary education             1   
15381         2         38    bachelor's degree             0   
14170         0         47  secondary education             1   
19117         1         25    bachelor's degree             0   
3052          0         51    bachelor's degree             0   
3524          0         54  secondary education             1   
3001          0         28    bachelor's degree             0   
17447         0         43  secondary education             1   
12369         0         58  secondary education             1   
16978         0         39    primary education             3   
19385         0         46  secondary education             1   
16215         0         65  secondary education             1   
9929          0         52    bachelor's degree             0   
20930         0         55  secondary education             1   
16998         0         51  secondary education             1   
15688         2         38  secondary education             1   
15431         0         28    bachelor's degree             0   
16333         1         49    bachelor's degree             0   

           family_status  family_status_i

In [75]:
credit_scoring['purpose_cat'].describe()

count      21465
unique         4
top       estate
freq       10810
Name: purpose_cat, dtype: object

Here I categorized the data by the loan purpose. There are 38 unique purpose entries, every single entry contain one of the following words: 'car', 'estate', 'housing', 'property', 'house', 'education', 'educated', 'university', 'wedding'. So it was easy to create function sorting the data according to the words strings contain. 

###### 2.4.2 categorization by income type

In [76]:
def income_cat(row, first, second, third, fourth):
    income_cat = ''
    
    
    try:  
        
        if row['total_income'] <= first:
            income_cat = 'low'
        elif first <= row['total_income'] < second:
            income_cat = 'lower-middle'
        elif second <= row['total_income'] < third:
            income_cat = 'upper-middle'
        elif third <= row['total_income'] < fourth:
            income_cat = 'high'
                 
        return income_cat    
        
    except: 
        
        print('Oy vey! Smthng went wrong')
    
    
first_quarter = credit_scoring.total_income.quantile(0.25)
second_quarter = credit_scoring.total_income.quantile(0.5)
third_quarter = credit_scoring.total_income.quantile(0.75)
fourth_quarter =  credit_scoring.total_income.quantile(1)


credit_scoring['income_cat'] = credit_scoring.apply(income_cat, args=(first_quarter, second_quarter, third_quarter, fourth_quarter), axis=1)


credit_scoring.sample(60)

children  dob_years            education  education_id  \
14572         0         47    bachelor's degree             0   
161           1         39    bachelor's degree             0   
15072         1         39  secondary education             1   
20753         1         44  secondary education             1   
12104         2         38  secondary education             1   
9009          0         48  secondary education             1   
17358         1         45    bachelor's degree             0   
9815          0         41  secondary education             1   
13212         0         46  secondary education             1   
6298          0         57  secondary education             1   
14139         1         45  secondary education             1   
11733         0         19  secondary education             1   
15673         2         51    bachelor's degree             0   
19978         0         61  secondary education             1   
1010          0         66  secondary education             1   
18121         0         36  secondary education             1   
8183          0         60  secondary education             1   
5926          0         29  secondary education             1   
20572         0         52  secondary education             1   
1336          0         32    bachelor's degree             0   
8991          1         48         some college             2   
19787         1         43  secondary education             1   
6998          0         48  secondary education             1   
1425          0         55  secondary education             1   
17921         0         24  secondary education             1   
21151         2         37    bachelor's degree             0   
13868         1         50    bachelor's degree             0   
9450          2         37  secondary education             1   
5579          1         42  secondary education             1   
20941         0         57  secondary education             1   
13902         0         44         some college             2   
3925          0         60    bachelor's degree             0   
15803         0         44  secondary education             1   
19193         1         30    bachelor's degree             0   
8354          0         56  secondary education             1   
13564         0         58  secondary education             1   
6345          0         52  secondary education             1   
14804         0         60  secondary education             1   
2104          1         63  secondary education             1   
3367          2         35         some college             2   
13259         0         25  secondary education             1   
7070          1         42    bachelor's degree             0   
8169          0         60  secondary education             1   
145           0         62  secondary education             1   
15443         0         23  secondary education             1   
9081          0         24    bachelor's degree             0   
13600         0         64  secondary education             1   
17984         2         31  secondary education             1   
5619          1         41    bachelor's degree             0   
1882          0         40  secondary education             1   
15553         1         34    primary education             3   
16661         0         60    bachelor's degree             0   
3864          1         45  secondary education             1   
13661         3         37  secondary education             1   
2268          2         34  secondary education             1   
13275         0         55  secondary education             1   
461           0         36  secondary education             1   
17659         1         45    primary education             3   
8485          0         59  secondary education             1   
17255         0         56  secondary education             1   

           family_status  family_status_i

Not sure how try-except method is actually useful here, I added just to be, saw it's funny error message several times. What an elegant solution with the quantile() function, just discovered it thanks to my cohortmate. Here income_cat function defined to sort entries into four groups, depending on which quantile their total_income value belongs to. 

### Step 3. Answering questions

#### 3.1 Is there a connection between having kids and repaying a loan on time?

In [77]:
credit_scoring['children'].value_counts()

0    14226
1     4808
2     2051
3      330
4       41
5        9
Name: children, dtype: int64

In [78]:
credit_scoring['debt'].value_counts()

0    19726
1     1739
Name: debt, dtype: int64

Let's define a pivot table displaying default rate by number of children:

In [79]:
debt_by_kids = pd.pivot_table(credit_scoring, values=['debt'],  index=['children'], aggfunc = sum)
debt_by_kids['total_count'] = credit_scoring['children'].value_counts()#.astype(int)
debt_by_kids['default_rate_%'] =((debt_by_kids['debt'] / debt_by_kids['total_count']) * 100).round(2)
debt_by_kids.sort_values(['default_rate_%'], ascending = False)

debt  total_count  default_rate_%
children                                   
4            4           41            9.76
2          193         2051            9.41
1          443         4808            9.21
3           27          330            8.18
0         1072        14226            7.54
5            0            9            0.00

##### Conclusion

So here I arranged data into a pivot table and showed the default rate for groups with different number of children. 

Seems like people with 4 children are the most prone to default loans with 9.76% default rate. People having 2 and 1 kids are going closely after with 9.41 and 9.21 percents respectively. Customers having 3 kids have noticeable less rate of 8.18%, and those with no kids are more credible with 7.54% rate according to our data. Persons with 5 children look like the most credible with 0 default rate, but it may be a result of their underrepresentation with only 9 entries in our dataset, so I'm considering their result as insignificant here. 

#### 3.2 Is there a connection between marital status and repaying a loan on time?

In [80]:
credit_scoring['family_status'].value_counts()

married              12342
civil partnership     4160
unmarried             2809
divorced              1195
widow / widower        959
Name: family_status, dtype: int64

Here's a pivot table showing default rate for customers of diferent family status:

In [81]:
debt_by_martstat = pd.pivot_table(credit_scoring, values=['debt'],  index=['family_status'], aggfunc = sum)
debt_by_martstat['total_count'] = credit_scoring['family_status'].value_counts()
debt_by_martstat['default_rate_%'] =((debt_by_martstat['debt'] / debt_by_martstat['total_count']) * 100).round(2)
debt_by_martstat.sort_values(['default_rate_%'], ascending = False)

debt  total_count  default_rate_%
family_status                                       
unmarried           274         2809            9.75
civil partnership   388         4160            9.33
married             929        12342            7.53
divorced             85         1195            7.11
widow / widower      63          959            6.57

##### Conclusion

Here I see that unmarried customers have the highest default rate of 9.75%, the civil partners are next with 9.33%, married people are significantly more reliable, rated 7.53%. Divorced and widowed persons have the least rates of 7.11% and 6.57 respectively.

#### 3.3 Is there a connection between income level and repaying a loan on time?

This is a pivot to see the default rate for different income groups:

In [82]:
debt_by_inc = pd.pivot_table(credit_scoring, values=['debt'],  index=['income_cat'], aggfunc = sum)
debt_by_inc['total_count'] = credit_scoring['income_cat'].value_counts()
debt_by_inc['default_rate_%'] =((debt_by_inc['debt'] / debt_by_inc['total_count']) * 100).round(2)
debt_by_inc.sort_values(['default_rate_%'], ascending = False)

debt  total_count  default_rate_%
income_cat                                     
lower-middle   377         4306            8.76
upper-middle   554         6425            8.62
low            425         5367            7.92
high           383         5366            7.14
                 0            1            0.00

##### Conclusion

lower-middle quarter income customers are inclined to default loans the most here, 8.76%, with upper-middle quarter closely behind, 8.62%. Low income people are slightly more reliable with 7.92%. High income customers, unsurprisingly, are the most reliable with 7.14% rate.

#### 3.4 How do different loan purposes affect on-time loan repayment?

Pivot of defaulting rates by loan purposes: 

In [83]:
debt_by_purp = pd.pivot_table(credit_scoring, values=['debt'],  index=['purpose_cat'], aggfunc = sum)
debt_by_purp['total_count'] = credit_scoring['purpose_cat'].value_counts()
debt_by_purp['default_rate_%'] =((debt_by_purp['debt'] / debt_by_purp['total_count']) * 100).round(2)
debt_by_purp.sort_values(['default_rate_%'], ascending = False)

debt  total_count  default_rate_%
purpose_cat                                   
vehicle       402         4307            9.33
education     370         4014            9.22
wedding       186         2334            7.97
estate        781        10810            7.22

##### Conclusion

Looks like vehicle buyers are the least reliable debtors with the default rate of 9.33%, students are very close with 9.22%. Newlyweds have significantly more reliable 7.97% rate, and estate investors have the best rate of 7.22%.

### Step 4. General conclusion

All done, checklist filled completely. Initial dataset was corrupted heavily, but I managed to fix it.

Yes, customer's marital status and number of children do has significant impact on their chances to default on a loan.

Here I found that the high income, divorsed and widowed, estate investors, and childfree customers are the most reliable groups of potential debtors, while lower-middle quarter income, unmarried, and those having 4 kids are the least reliable fractions according to the dataset.

Unmarried customers have the highest default rate of 9.75%, the civil partners are next with 9.33%, married people are significantly more reliable, rated 7.53%. Divorced and widowed persons have the least rates of 7.11% and 6.57 respectively.

People with 4 children are the most prone to default loans with 9.76% default rate. People having 2 and 1 kids are going closely after with 9.41 and 9.21 percents respectively. Customers having 3 kids have noticeable less rate of 8.18%, and those with no kids are more credible with 7.54% rate according to our data.

Customers with 5 kids excluded from the final report fot being underrepresented. Default rates distribution look counterintuitive to me at some points, but that's what make it even more exciting. 